src/main/scala/uk/gov/nationalarchives/Processor.scala

```diff
@@ -29,52 +30,139 @@ class Processor(
+
+    for {
+      contentObjectsFromRep <- entityClient.getContentObjectsFromRepresentation(
+        ioRef,
+        representationType,
+        index
+      )
+    } yield contentObjectsFromRep.collect {
+      case contentObjectFromRep if contentObjectFromRep.ref == coRef => s"${representationType}_$index"
+    }
+  }
+
+  private def createDestinationFilePath(
+      ioRef: UUID,
+      potentialRef: Option[UUID] = None,
+      potentialRepTypeGroup: Option[String] = None,
+      potentialGenType: Option[GenerationType] = None,
+      potentialGenVersion: Option[Int] = None,
+      fileName: String
+  ): String =
+    List(
+      Some(ioRef),
+      potentialRepTypeGroup,
+      potentialRef,
+      potentialGenType.map(_.toString.toLowerCase),
+      potentialGenVersion.map(version => s"g$version"),
+      Some(fileName)
+    ).flatten.mkString("/")
+
+  private def createMetadataFileName(entityType: String) = s"${entityType}_Metadata.xml"
+
   private def toDisasterRecoveryObject(message: Message): IO[List[DisasterRecoveryObject]] = message match {
     case InformationObjectMessage(ref, _) =>
       for {
         entity <- fromType[IO](EntityClient.InformationObject.entityTypeShort, ref, None, None, deleted = false)
-        metadata <- entityClient.metadataForEntity(entity).map { metadata =>
-          createMetadataObject(entity, metadata)
+        metadataFileName = createMetadataFileName(entity.entityType.get.entityTypeShort)
+        metadataObject <- entityClient.metadataForEntity(entity).map { metadata =>
+          val destinationFilePath = createDestinationFilePath(entity.ref, fileName = metadataFileName)
+          createMetadataObject(entity.ref, metadata, metadataFileName, destinationFilePath)
         }
-      } yield metadata
+      } yield metadataObject
     case ContentObjectMessage(ref, _) =>
       for {
         bitstreamInfoPerCo <- entityClient.getBitstreamInfo(ref)
-        entity <- entityClient.getEntity(ref, ContentObject)
+        entity <- fromType[IO](
+          EntityClient.ContentObject.entityTypeShort,
+          ref,
+          None,
+          None,
+          deleted = false,
+          parent = bitstreamInfoPerCo.headOption.flatMap(_.parentRef)
+        )
         parentRef <- IO.fromOption(entity.parent)(new Exception("Cannot get IO reference from CO"))
-      } yield bitstreamInfoPerCo.toList.map(bitStreamInfo =>
-        FileObject(parentRef, bitStreamInfo.name, bitStreamInfo.fixity.value, bitStreamInfo.url)
-      )
+        urlsOfRepresentations <- entityClient.getUrlsToIoRepresentations(parentRef, None)
+        coRepTypes <- urlsOfRepresentations.map(getRepresentationTypeOfCo(parentRef, _, entity.ref)).flatSequence
+        _ <- IO.raiseWhen(coRepTypes.length > 1) {
+          new Exception(s"${entity.ref} belongs to more than 1 representation type: ${coRepTypes.mkString(", ")}")
+        }
+        representationTypeGroup = coRepTypes.head
```

Comment:
If you do `headOption` here then you don't have to wrap it in `Some` further down.